keel-igor/src/main/kotlin/com/netflix/spinnaker/keel/services/DeliveryConfigImporter.kt

```diff
@@ -0,0 +1,58 @@
+package com.netflix.spinnaker.keel.services
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.convertValue
+import com.netflix.spinnaker.igor.ScmService
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.core.api.SubmittedDeliveryConfig
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.keel.services.AuthorizationSupport.Action.WRITE
+import com.netflix.spinnaker.keel.services.AuthorizationSupport.TargetEntity.APPLICATION
+import com.netflix.spinnaker.keel.services.AuthorizationSupport.TargetEntity.SERVICE_ACCOUNT
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+
+/**
+ * Provides functionality to import delivery config manifests from source control repositories (via igor).
+ */
+@Component
+class DeliveryConfigImporter(
+  private val jsonMapper: ObjectMapper,
+  private val scmService: ScmService,
+  private val keelRepository: KeelRepository,
+  private val authorizationSupport: AuthorizationSupport
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  /**
+   * Imports a delivery config from source control via igor.
+   */
+  fun import(
+    repoType: String,
+    projectKey: String,
+    repoSlug: String,
+    manifestPath: String,
+    ref: String
+  ): DeliveryConfig {
+    val manifestLocation = "$repoType://project:$projectKey/repo:$repoSlug/manifest:$manifestPath@$ref"
+
+    log.debug("Retrieving delivery config from $manifestLocation")
+    val submittedDeliveryConfig = runBlocking {
+      scmService.getDeliveryConfigManifest(repoType, projectKey, repoSlug, manifestPath, ref)
+        .let {
+          jsonMapper.convertValue<SubmittedDeliveryConfig>(it)
+        }
+    }
+
+    log.debug("Successfully retrieved delivery config from $manifestLocation. Checking permissions to upsert...")
+    authorizationSupport.checkApplicationPermission(WRITE, APPLICATION, submittedDeliveryConfig.application)
+    authorizationSupport.checkServiceAccountAccess(SERVICE_ACCOUNT, submittedDeliveryConfig.serviceAccount)
+
+    log.debug("Upserting delivery config '${submittedDeliveryConfig.name}' for app '${submittedDeliveryConfig.application}' from $manifestLocation")
+    val deliveryConfig = keelRepository.upsertDeliveryConfig(submittedDeliveryConfig)
```

Comment:
can you please add `validation` here as well?

Reply:
Oh, good catch. I assumed `keelRepository.upsertDeliveryConfig` called the validator but I think that was before your change, right?

Reply:
nope, it's decoupled from `keelRepository`.

Reply:
Done!